### ISSUE TITLE
Fatal error: exception Lib.Promise.Promise

### ISSUE THREAD

**Issue by FloorLamp** [dfinity/motoko#2714]
I hit a compiler bug trying to upgrade some data structures to a new version.

Diff: https://github.com/FloorLamp/axon/commit/672d871d574e751d08285f109a64cf4c349318fb

```
OOPS! You've triggered a compiler bug.
Please report this at https://github.com/dfinity/motoko/issues/new with the following details:

Motoko (source dphsbbg2-hx7ajr94-7j7higa7-v6hfzw6m)

Fatal error: exception Lib.Promise.Promise
Raised at Lib.Promise.fulfill in file "lib/lib.ml", line 456, characters 55-68
Called from Lib.Promise.lazy_value in file "lib/lib.ml", line 461, characters 34-50
Called from Codegen__Compile.StackRep.adjust in file "codegen/compile.ml", line 5316, characters 48-75
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7566, characters 14-46
Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23
Called from Codegen__InstrList.concat_map in file "codegen/instrList.ml" (inlined), line 135, characters 43-58
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6866, characters 6-50
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6812, characters 10-61
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6813, characters 24-34
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6848, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.StackRep.materialize_const_v in file "codegen/compile.ml", line 5283, characters 53-64
Called from Lib.Promise.lazy_value in file "lib/lib.ml", line 461, characters 44-50
Called from Codegen__Compile.StackRep.adjust in file "codegen/compile.ml", line 5316, characters 48-75
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7566, characters 14-46
Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23
Called from Codegen__InstrList.concat_map in file "codegen/instrList.ml" (inlined), line 135, characters 43-58
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6866, characters 6-50
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6812, characters 10-61
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.StackRep.materialize_const_v in file "codegen/compile.ml", line 5283, characters 53-64
Called from Lib.Promise.lazy_value in file "lib/lib.ml", line 461, characters 44-50
Called from Codegen__Compile.StackRep.adjust in file "codegen/compile.ml", line 5316, characters 48-75
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7566, characters 14-46
Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23
Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
Called from Codegen__InstrList.concat_map in file "codegen/instrList.ml" (inlined), line 135, characters 43-58
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6866, characters 6-50
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6812, characters 10-61
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7492, characters 30-61
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 7495, characters 18-27
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.StackRep.materialize_const_v in file "codegen/compile.ml", line 5283, characters 53-64
Called from Lib.Promise.lazy_value in file "lib/lib.ml", line 461, characters 44-50
Called from Codegen__Compile.StackRep.adjust in file "codegen/compile.ml", line 5316, characters 48-75
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7566, characters 14-46
Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23
Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
Called from Codegen__InstrList.concat_map in file "codegen/instrList.ml" (inlined), line 135, characters 43-58
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6866, characters 6-50
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6812, characters 10-61
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7492, characters 30-61
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 7495, characters 18-27
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.StackRep.materialize_const_v in file "codegen/compile.ml", line 5283, characters 53-64
Called from Lib.Promise.lazy_value in file "lib/lib.ml", line 461, characters 44-50
Called from Codegen__Compile.StackRep.adjust in file "codegen/compile.ml", line 5316, characters 48-75
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7566, characters 14-46
Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23
Called from Codegen__InstrList.concat_map in file "codegen/instrList.ml" (inlined), line 135, characters 43-58
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6866, characters 6-50
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6812, characters 10-61
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6813, characters 24-34
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6848, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.StackRep.materialize_const_v in file "codegen/compile.ml", line 5283, characters 53-64
Called from Lib.Promise.lazy_value in file "lib/lib.ml", line 461, characters 44-50
Called from Codegen__Compile.StackRep.adjust in file "codegen/compile.ml", line 5316, characters 48-75
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7566, characters 14-46
Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23
Called from Codegen__InstrList.concat_map in file "codegen/instrList.ml" (inlined), line 135, characters 43-58
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6866, characters 6-50
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6812, characters 10-61
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.StackRep.materialize_const_v in file "codegen/compile.ml", line 5283, characters 53-64
Called from Lib.Promise.lazy_value in file "lib/lib.ml", line 461, characters 44-50
Called from Codegen__Compile.StackRep.adjust in file "codegen/compile.ml", line 5316, characters 48-75
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7566, characters 14-46
Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23
Called from Codegen__InstrList.concat_map in file "codegen/instrList.ml" (inlined), line 135, characters 43-58
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6866, characters 6-50
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6812, characters 10-61
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6813, characters 24-34
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6848, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.StackRep.materialize_const_v in file "codegen/compile.ml", line 5283, characters 53-64
Called from Lib.Promise.lazy_value in file "lib/lib.ml", line 461, characters 44-50
Called from Codegen__Compile.StackRep.adjust in file "codegen/compile.ml", line 5316, characters 48-75
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7566, characters 14-46
Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23
Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39
Called from Codegen__InstrList.concat_map in file "codegen/instrList.ml" (inlined), line 135, characters 43-58
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6866, characters 6-50
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6812, characters 10-61
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7492, characters 30-61
Called from Codegen__Compile.compile_exp.go in file "codegen/compile.ml", line 7493, characters 17-28
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 7495, characters 18-27
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6813, characters 24-34
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6848, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.compile_exp in file "codegen/compile.ml", line 6849, characters 6-36
Called from Codegen__Compile.compile_exp_as in file "codegen/compile.ml", line 7565, characters 24-44
Called from Codegen__Compile.FuncDec.compile_local_function.(fun) in file "codegen/compile.ml", line 5583, characters 20-37
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.def in file "lib/lib.ml", line 499, characters 13-28
Called from Codegen__Compile.compile_dec.(fun) in file "codegen/compile.ml", line 7833, characters 39-50
Called from Codegen__Compile.compile_dec.(fun) in file "codegen/compile.ml", line 7818, characters 30-42
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.go.(fun) in file "codegen/compile.ml", line 7866, characters 44-62
Called from Codegen__Compile.compile_decs_public.(fun) in file "codegen/compile.ml", line 7869, characters 39-61
Called from Codegen__Compile.main_actor.(fun) in file "codegen/compile.ml", line 8064, characters 4-21
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27
Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11
Called from Lib.AllocOnUse.use in file "lib/lib.ml", line 514, characters 13-28
Called from Codegen__Compile.E.built_in in file "codegen/compile.ml" (inlined), line 401, characters 4-49
Called from Codegen__Compile.IC.export_init.(fun) in file "codegen/compile.ml", line 3453, characters 20-43
Called from Codegen__Compile.Func.of_body in file "codegen/compile.ml", line 685, characters 6-18
Called from Codegen__Compile.IC.export_init in file "codegen/compile.ml", line 3450, characters 18-227
Called from Codegen__Compile.compile in file "codegen/compile.ml", line 8167, characters 6-24
Called from Pipeline.compile_files in file "pipeline/pipeline.ml", line 647, characters 14-53
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Diag.bind in file "lang_utils/diag.ml", line 32, characters 27-30
Called from Moc.process_files in file "exes/moc.ml", line 148, characters 37-82
Called from Moc in file "exes/moc.ml", line 210, characters 4-23
```

**Comment by nomeata** [comment]
(This would be easier to debug if we’d be allowed annotate promises with debug information, even if it is only ever used when our code is buggy (as it is here). </rant>)

I can reproduce it with with current master with
```
cd /tmp
git clone https://github.com/FloorLamp/axon
cd axon/
git checkout 672d871d574e751d08285f109a64cf4c349318fb
cd ~/dfinity/motoko
moc --package base ../motoko-base/src/ /tmp/axon/src/Axon/Axon.mo 
```

**Comment by nomeata** [comment]
Hmm, the error appears to be related to our handling of constant static values. Maybe some recursive definitions that the compiler is trying to emit in the “wrong” order?

**Comment by nomeata** [comment]
Hmm, can’t do more debugging right now. Probably best to try to minimize the diff or the reproducer.

**Comment by FloorLamp** [comment]
Minimal repro:

```
actor Repro {
  func multiple_trycatch() : async () {
    try {
      await foo();
    } catch (error) {};

    try {
      await foo();
    } catch (error) {};
  };

  shared func foo() : async () {};
};
```

**Comment by FloorLamp** [comment]
Here is a minimal repro: https://github.com/FloorLamp/axon/commit/1c8b1c417180d29739b35b8d70856755be3b973e

**Comment by ggreif** [comment]
Hey @FloorLamp thanks for the repro. I think I can debug this.

**Comment by nomeata** [comment]
Thanks! Smaller, self-contained repro:
```
actor Repro {
  public type A = { x : ??{ #y : A }; };
  public shared func foo(blob: ?A) : async () {
    ignore debug_show(blob)
  };
};
```

**Comment by nomeata** [comment]
Fun fact: Adding the `bar` function as follows, even unused, makes the problem go away:
```
type A = { #y : A };
func bar(blob: ?A) : Text {
  switch (blob) {
    case (?x) { debug_show(x) };
    case null "null";
  };
};
func foo(blob: ?A) : () {
  ignore (debug_show(blob));
};
foo(null);
```
Very curious.

**Comment by nomeata** [comment]
Gotta run now, but I sense some bad interaction with the laziness in the handling of constant data: Under certain (not fully understood) circumstances, a recursive top-level function will upon its first use call `materialize_const_t`, which starts filling the promise that will eventually contain the static memory location of that object. This kicks off `materialize_const_v`, which calculates the memory content. Normally that is done quickly, but with a function (`Const.Func`) I suspect that it can now try to read the promise before this is done.

Maybe the contents of `Const.Func` should not just be `get_fn : unit -> int32`, but rather something two-step (`unit -> (int32 * unit -> unit)`) so that we can fill the promise with the function number faster, and _then_ actually compile the function.

But it’s not quite pleasing this way, maybe there is a more elegant way to solve this. Or maybe some appropriate abstraction can help.

**Comment by nomeata** [comment]
Ok, I have not just one fix, but two (#2716 and #2717). Not sure yet which one I prefer. Any opinions, @ggreif or @crusso?
